go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,269 @@
+# Go SDK Workflow Interceptor Updates and OpenTelemetry Tracing Support
```

Comment:
We should mention that we plan to break compatibility for current generation of interceptors here.
Also, this pretty much deprecates context propagators and built-in tracing support.
I'd share this spec with the community via the `#go-sdk` slack channel.

Reply:
Built-in tracing could be implemented as an interceptor.
Should we try to do that instead of going for opentelemetry?

Reply:
> There is a big bold note that says:
> These updates are known to NOT be backwards-compatible

Missed that :/

Reply:
We currently have 3 different tracing concepts.
1. built-in tracing in golang
2. opentracing interceptors in java
3. opentelemetry interceptors in node

The Java and Node interceptors are incompatible.

Java uses this header key and span context definition:
- https://github.com/temporalio/sdk-java/blob/6f86b160d9b948418d2e49d8e5a2bbb260fad149/temporal-opentracing/src/main/java/io/temporal/opentracing/internal/ContextAccessor.java#L36
- https://github.com/opentracing/opentracing-java/blob/a8557d7c7a56210cbcf6bc518fe24b32923aa66f/opentracing-api/src/main/java/io/opentracing/SpanContext.java

Node uses this header key and span context definition:
- https://github.com/temporalio/sdk-node/blob/4505eee94e7d8a10bc187612977fd72bc6d740a6/packages/common/src/otel.ts#L5
- https://github.com/open-telemetry/opentelemetry-js-api/blob/41109c83a9784d689f319f2c5d953b3874c694a3/src/trace/span_context.ts#L23

I'm not sure what go uses but ideally all of our SDKs will be compatible.
We should for now keep the go built-in tracing support backwards compatible.

3 options I see:
- Re-implement built-in tracing using interceptors
- Create a new opentelemetry / opentracing interceptors package that's compatible with at least Java or Node
- Do both

Reply:
Java has both opentelementry and opentracing. Opentelemetry is easily pluggable through opentelementry-shim and fully functional.
https://github.com/temporalio/sdk-java/tree/master/temporal-opentracing#opentelemetry